libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/DefaultTracer.java

```diff
@@ -77,7 +77,11 @@ public SpanContext getCurrentSpan() {
     private void endSpan(Span span) {
         if (span != null) {
             span.endSpan();
-            setCurrentSpanInContext(span.getParentSpan());
+            if (span.getParentSpan() != null && !span.getParentSpan().hasEnded()) {
```

Comment:
@Gaganjuneja the thing you're trying to fix should never happen, I think we are managing context restoration incorrectly. If the execution is synchronous, thing are clear: the parent could not be finished before children, not the asynchronous case is different:
 - the parent span and child spans do not belong to the same thread
 - when the we start new span scope, we should capture the current span associated with the current thread (which could be not the same as the parent span)
 - when the span is ended, we have to restore the span that was active on that particular thread

Reply:
Thanks @reta for taking a look.
I had run a couple of tests and deep dive to understand the issue.

As of now Our context restoration in based on the fact that https://github.com/opensearch-project/OpenSearch/blob/8cfde6ca4031047b4f7e5ecf706e434e1c535567/server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java#L74 `Every opensearch thread is managed by a thread pool or executor being responsible for stashing and restoring the threads context`. So ideally even we are setting the wrong/ended parent but it should be removed once the thread call is over and context is restored. One specific scenario where this is not happening is scheduled cluster monitoring actions.

<img width="602" alt="Screenshot 2023-08-27 at 10 08 16 AM" src="https://github.com/opensearch-project/OpenSearch/assets/2391016/b9381ebf-e1fb-47eb-b178-c17a5d80e673">

Here if at Thread1 the context is not stashed the state will be persisted. This exactly happens with the NodeStatsTransportAction and possibly with other actions where startSpan through TransportService::sendRequest is creating the span and persisting with the current Management thread without stashing the context after the call is over.

We can fix this by adding the ThreadContext.stashContext() here https://github.com/opensearch-project/OpenSearch/pull/9415/files#diff-518cfddb0699fc99133f7a7dd5dc2224465c903405ae4aa9a8440abd03d1a686R868

We can fix this case by case. But now the question is should we add a check at the framework level like when we are starting the span then check if the current span is already ended then ignore that and create a new span without the current parent. Capturing the currentSpan while creating the span also wont help because most of the cases the context is being propagated to a caller thread from the calling thread.

Your thoughts on this please.

Reply:
There could be more cases from the ClusterApplierService.

Reply:
Thanks for looking @Gaganjuneja , so indeed we don't manage context properly. So we need to make very clear distinction between spans and scopes. The `OpenTelemetry` models it differently but we don't: the spans could outlive the scopes, but the scopes provide this short-living context propagation (I will share the idea and will help you to deal with API changes):

```
            // Creates a Span that is not associated with the thread (it is not attached to anything)
            final Span httpRequestSpan = tracer.startSpan(SpanBuilder.from(httpRequest));
            // Attaches the span to current thread (scope)
            final (final SpanScope scope = tracer.spanInScope(httpRequestSpan)) {
                HttpChannel traceableHttpChannel = new TraceableHttpChannel(httpChannel, httpRequestSpanScope);
                handleIncomingRequest(httpRequest, traceableHttpChannel, httpRequest.getInboundException());
            } 
            // When scope is closed, the previously attached span is restored, but the current span is not closed
```

With this model (span + scope), we will be able to cover any async and reactive context propagation. On the side note, OpenTelemetry has pretty good debugging capabilities to help validating the spans / scopes lifecycle:

```
-Dio.opentelemetry.context.enableStrictContext=true
```

We are building our MockXxx implementation but I think we should stop doing it and use OpenTelemetry in tests instead, it will save us tons of time and efforts.